### ISSUE TITLE
Support AVX-512

### ISSUE THREAD

**Issue by lemire** [simdjson/simdjson#10]
This is currently a low priority but it seems worthwhile to ask whether AVX-512 helps, and by how much. Of course, the work should be completed on AVX2 first.

**Comment by lemire** [comment]
@fwessels reports good results using AVX-512 in the Go port of simdjson...

https://github.com/minio/simdjson-go#avx512-acceleration


cc @jkeiser

**Comment by jkeiser** [comment]
Nice! When I sat down and tried to get it in, it wasn't as straightforward as just adding a new simd8 implementation, just because the avx-512 mask registers are so different from simd FF masks. Haven't revisited for a couple of months.

**Comment by fwessels** [comment]
As a tip: you might want to disassemble the AVX512 instructions since the Golang AVX512 instructions follow to "Plan9/AT&T" syntax (see https://github.com/golang/go/wiki/AVX512 for more details).

**Comment by lemire** [comment]
@jkeiser I think that what @fwessels does is to systematically convert the AVX-512 masks to general purpose registers (KMOVQ):


```
TEXT ·__find_whitespace_and_structurals_avx512(SB), $0
    LEAQ LCDATA1<>(SB), BP

    VMOVDQA32   Z8, Z0
    VMOVDQU32   (BP), Z2
    VPSHUFB     Z0, Z2, Z3
    VPSRLD      $4, Z0, Z0
    VMOVDQU32   0x40(BP), Z4
    VPANDD      Z4, Z0, Z0
    VMOVDQU32   0x80(BP), Z5
    VPSHUFB     Z0, Z5, Z0
    VPANDD      Z3, Z0, Z0
    VMOVDQU32   0xc0(BP), Z2
    VPANDD      Z2, Z0, Z3
    VPXORD      Z4, Z4, Z4
    VPCMPEQB    Z4, Z3, K1
    KMOVQ       K1, SI
    NOTQ        SI
    MOVQ        SI, (CX)
    VMOVDQU32   0x100(BP), Z2
    VPANDD      Z2, Z0, Z0
    VPCMPEQB    Z4, Z0, K1
    KMOVQ       K1, CX
    NOTQ        CX
    MOVQ        CX, (DX)
    RET
```
https://github.com/minio/simdjson-go/blob/master/find_whitespace_and_structurals_amd64.s

So I think no magic is necessary. 

See the comment above where a straight-forward AVX-512 gave me "maybe 5%". Though @fwessels does better, he surely did a finer job than my hasty port. So 10% to 15% should be achievable with a few days of work.

*However* AVX-512 has other benefits that should not be underestimated. For example, `vpcompressb` is likely to bring a massive gain in some of the code.

And that's really the end game. A straight-forward port may only achieve 10%, say... but once you have that, you can start exploring all of the *new* instructions and that's where it is going to really shine.

Here is some recommended reading ....


https://arxiv.org/abs/1910.05109

**Comment by fwessels** [comment]
Correct, this port is a relatively "dumb" (or straightforward) port from AVX2 to AVX512 for the various Stage 1 algorithms, typically halving the number of SIMD instructions (2 Y-registers are needed for AVX2 whereas a single Z register is sufficient for AVX512). 

For sure there should be room for significantly more performance improvements by tuning the algorithms specifically for AVX512. And that is where the fun really starts ....

**Comment by jkeiser** [comment]
Oh, I see. Hmm, one might be able to implement simd8<bool> as a mask register and expand to ff simd registers when operators require it. Not perfect but still probably pretty good.

In the limit, either the abstractions need to be updated or avx 512 needs to be a separate implementation. My gut says that we're going to need to change the algorithms in big enough ways that we'll want to split it. But for now I bet this could be made to work.

**Comment by jkeiser** [comment]
In the short term we could also make a mask64 type that is mmask on avx512 uint64_t for other architectures. I bet we can avoid some moves to/from gp 64 bit registers that way.

**Comment by jkeiser** [comment]
It's a good question: do we feel avx 512 blocks a 1.0? Would we feel embarrassed and lose some real traction if the new shiny simd json library doesn't support the latest greatest shiny simd?

**Comment by fwessels** [comment]
> In the short term we could also make a mask64 type that is mmask on avx512 and uint64_t for other architectures. I bet we can avoid some moves to/from gp 64 bit registers in the generic code that way.

I guess that would work, we can probably do more using the K registers (golang version as well), will take another look at the code, see if there is some low hanging fruit ...

**Comment by lemire** [comment]
No. AVX-512 does not block any release in my mind.

It is not a given that staying in mask form is better. In fact, it may be worse because, from my recollection, the operations on masks compete with other AVX-512 operations. I don't think many people have lots of experience with AVX-512 but my current belief is that if you have to do lots of work on the masks, you are better off converting them to general purpose registers. (I would need to qualify this statement.)

**Comment by jkeiser** [comment]
Oh! So it really might only be useful if you are only doing a small number of operations, aren't using SIMD very heavily, or intend to use the mask again afterwards in a SIMD operation.

**Comment by fwessels** [comment]
>  It is not a given that staying in mask form is better. In fact, it may be worse because, from my recollection, the operations on masks compete with other AVX-512 operations. I don't think many people have lots of experience with AVX-512 but my current belief is that if you have to do lots of work on the masks, you are better off converting them to general purpose registers. (I would need to qualify this statement.)

Good question, don't know the answer. Perhaps if you can stick to K registers as much as possible it may be faster overall as compared to moving them to GPRs and continuing there. It would be interesting to get some "hard" data on this ...

**Comment by jkeiser** [comment]
I may try the approach but apply it incrementally and see which places help and which hurt ...

**Comment by lemire** [comment]
>  Perhaps if you can stick to K registers as much as possible it may be faster overall as compared to moving them to GPRs and continuing there.

So if you have the mask register and need to convert it, you pay for it dearly. I think it can be several cycles of latency (say 3?). But then, after that, you have cheap general-purpose arithmetic.

If you stick with masks, you avoid the expensive conversions. However, you are stuck with the "k*" instructions that are much more expensive.

On a modern x64 system, you can do 4 bitwise AND in one cycle. But you can only do one kand per cycle. And I think it interferes with your ability to do SIMD instructions at the same time (same execution units).

So if you are doing lots of operations on your masks, I think you should convert them.

In other words, I don't think that @fwessels  is going to find that he has much to gain by avoiding the kmov. Now, I could be wrong, and it is worth checking my speculations with hard measurements.

**Comment by jkeiser** [comment]
I played a little bit with it, and the smallest implementation of `prev` (lane-crossing alignr_epi8) I could find was this: (i.e. `_mm512_alignr_epi8(_mm512_alignr_epi64(prev, current, 1), 16 - N)`).

**Comment by lemire** [comment]
@jkeiser Did you check out how I did it?

https://github.com/lemire/fastvalidate-utf-8/blob/master/include/simdutf8check_avx512.h#L33

Thoughts?

**Comment by jkeiser** [comment]
@lemire I didn't! That's nifty. I'll dip into that reservoir more.

I was trying to restrict myself to immediate values, the 2 input registers, and the output register, and not add more constants in an already somewhat register-pressured situation. But Skylake has double the registers, so your solution will probably run faster and I probably need to get out of that habit :)

[Isn't `_mm512_permutex2var_epi8` AVX512_VBMI though?](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm512_permutex2var_epi8&expand=3876,3965,4262) [Wikipedia indicates](https://en.wikipedia.org/wiki/AVX-512#CPUs_with_AVX-512) that Skylake-X doesn't have VBMI, so I veered away from it when I was looking at permute instructions.

**Comment by lemire** [comment]
@jkeiser That's the fun thing with AVX-512... it is not one instruction set, but multiple instruction sets... I expect that Skylake-X will soon look obsolete and VBMI will be ubiquitous. It is such a powerful instruction set... for the kind of work we do, it will make a big difference.

I have a remotely accessible box with VBMI, I will email you.

**Comment by lemire** [comment]
We have had an AVX-512 kernel for quite some time now. Closing this.
